File: akka-samples/akka-sample-persistence-scala/src/main/scala/sample/persistence/ViewExample.scala
package sample.persistence

import scala.concurrent.duration._

import akka.actor._
import akka.persistence._

object ViewExample extends App {
  class ExamplePersistentActor extends PersistentActor {
    override def persistenceId = "sample-id-4"

    var count = 1

    def receiveCommand: Receive = {
      case payload: String =>
        println(s"persistentActor received ${payload} (nr = ${count})")
        persist(payload + count) { evt =>
          count += 1
        }
    }

    def receiveRecover: Receive = {
      case _: String => count += 1
    }
  }

  class ExampleView extends PersistentView {
    private var numReplicated = 0

    override def persistenceId: String = "sample-id-4"
    override def viewId = "sample-view-id-4"

    def receive = {
      case "snap" =>
        println(s"view saving snapshot")
        saveSnapshot(numReplicated)
      case SnapshotOffer(metadata, snapshot: Int) =>
        numReplicated = snapshot
        println(s"view received snapshot offer ${snapshot} (metadata = ${metadata})")
      case payload if isPersistent =>
        numReplicated += 1
        println(s"view replayed event ${payload} (num replicated = ${numReplicated})")
      case SaveSnapshotSuccess(metadata) =>
        println(s"view saved snapshot (metadata = ${metadata})")
      case SaveSnapshotFailure(metadata, reason) =>
        println(s"view snapshot failure (metadata = ${metadata}), caused by ${reason}")
      case payload =>
        println(s"view received other message ${payload}")
    }

  }

  val system = ActorSystem("example")

  val persistentActor = system.actorOf(Props(classOf[ExamplePersistentActor]))
  val view = system.actorOf(Props(classOf[ExampleView]))

  import system.dispatcher

  system.scheduler.schedule(Duration.Zero, 2.seconds, persistentActor, "scheduled")
  system.scheduler.schedule(Duration.Zero, 5.seconds, view, "snap")
}

File: akka-samples/akka-sample-persistence-java-lambda/src/main/java/sample/persistence/ViewExample.java
/**
 * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
 */

package sample.persistence;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;
import akka.persistence.*;
import scala.PartialFunction;
import scala.concurrent.duration.Duration;
import scala.runtime.BoxedUnit;

import java.util.concurrent.TimeUnit;

public class ViewExample {
  public static class ExamplePersistentActor extends AbstractPersistentActor {
    private int count = 1;

    @Override
    public String persistenceId() { return "sample-id-4"; }

    @Override
    public PartialFunction<Object, BoxedUnit> receiveCommand() {
      return ReceiveBuilder.
        match(String.class, s -> {
          System.out.println(String.format("persistentActor received %s (nr = %d)", s, count));
          persist(s + count, evt -> {
            count += 1;
          });
        }).
        build();
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receiveRecover() {
      return ReceiveBuilder.
        match(String.class, s -> count += 1).
        build();
    }
  }

  public static class ExampleView extends AbstractPersistentView {

    private int numReplicated = 0;

    @Override public String persistenceId() { return "sample-id-4"; }
    @Override public String viewId() { return "sample-view-id-4"; }

    public ExampleView() {
      receive(ReceiveBuilder.
        match(Object.class, m -> isPersistent(), msg -> {
          numReplicated += 1;
          System.out.println(String.format("view received %s (num replicated = %d)",
            msg,
            numReplicated));
        }).
        match(SnapshotOffer.class, so -> {
          numReplicated = (Integer) so.snapshot();
          System.out.println(String.format("view received snapshot offer %s (metadata = %s)",
            numReplicated,
            so.metadata()));
        }).
        match(String.class, s -> s.equals("snap"), s -> saveSnapshot(numReplicated)).build()
      );
    }
  }

  public static void main(String... args) throws Exception {
    final ActorSystem system = ActorSystem.create("example");
    final ActorRef persistentActor = system.actorOf(Props.create(ExamplePersistentActor.class));
    final ActorRef view = system.actorOf(Props.create(ExampleView.class));

    system.scheduler()
      .schedule(Duration.Zero(),
        Duration.create(2, TimeUnit.SECONDS),
        persistentActor,
        "scheduled",
        system.dispatcher(),
        null);
    system.scheduler()
      .schedule(Duration.Zero(), Duration.create(5, TimeUnit.SECONDS), view, "snap", system.dispatcher(), null);
  }
}
